test/api/model/submessage_test.dart

```diff
@@ -111,4 +116,176 @@ void main() {
+            ),
+        ]);
+    }
+
+    test('parse poll from message', () {
+      check(messageWithSubmessageContents(pollWidgetData))
+        .poll.isNotNull()
+          ..question.equals('favorite letter')
+          ..options.deepEquals(defaultOptions);
+    });
+
+    test('parse poll with new question event', () {
+      check(messageWithSubmessageContents(
+        pollWidgetData,
+        events: [
+          (eg.otherUser.userId, PollQuestionEventSubmessage(question: 'new question')),
+        ],
+      ))
+        .poll.isNotNull()
+          ..question.equals('new question')
+          ..options.deepEquals(defaultOptions);
+    });
+
+    test('parse poll with new option event', () {
+      check(messageWithSubmessageContents(
+        pollWidgetData,
+        events: [
+          (eg.otherUser.userId, PollNewOptionEventSubmessage(idx: 3, option: 'D')),
+          (eg.selfUser.userId, PollNewOptionEventSubmessage(idx: 0, option: 'E')),
+        ],
+      ))
+        .poll.isNotNull()
+          ..question.equals('favorite letter')
+          ..options.deepEquals([
+            ...defaultOptions,
+            PollOption(text: 'D'),
+            PollOption(text: 'E'),
+          ]);
+    });
+
+    test('parse poll with vote events on initial canned options', () {
+      check(messageWithSubmessageContents(
+        pollWidgetData,
+        events: [
+          (eg.otherUser.userId, PollVoteEventSubmessage(key: 'canned,1', op: PollVoteOp.add)),
+          (eg.otherUser.userId, PollVoteEventSubmessage(key: 'canned,2', op: PollVoteOp.add)),
+          (eg.otherUser.userId, PollVoteEventSubmessage(key: 'canned,2', op: PollVoteOp.remove)),
+          (eg.selfUser.userId, PollVoteEventSubmessage(key: 'canned,1', op: PollVoteOp.add)),
+        ],
+      ))
+        .poll.isNotNull()
+          ..question.equals('favorite letter')
+          ..options.deepEquals([
+            PollOption(text: 'A'),
+            eg.pollOption(text: 'B', voters: [eg.otherUser.userId, eg.selfUser.userId]),
+            PollOption(text: 'C'),
+          ]);
+    });
+
+    test('parse poll with vote events on post-creation options', () {
+      check(messageWithSubmessageContents(
+        pollWidgetData,
+        events: [
+          (eg.otherUser.userId, PollNewOptionEventSubmessage(idx: 0, option: 'D')),
+          (eg.otherUser.userId, PollVoteEventSubmessage(key: '${eg.otherUser.userId},0', op: PollVoteOp.add)),
+          (eg.otherUser.userId, PollVoteEventSubmessage(key: '${eg.selfUser.userId},0', op: PollVoteOp.add)),
+          (eg.otherUser.userId, PollVoteEventSubmessage(key: '${eg.selfUser.userId},0', op: PollVoteOp.remove)),
+        ],
+      ))
+        .poll.isNotNull()
+          ..question.equals('favorite letter')
+          ..options.deepEquals([
+            ...defaultOptions,
+            eg.pollOption(text: 'D', voters: [eg.otherUser.userId]),
+          ]);
+    });
+
+    test('applyEvent: adding repeated option', () {
```

Comment:
Having these test cases for error paths is great. Let's also have test cases for the normal happy paths, though. As a bonus, those can make the error tests clearer because they serve as a baseline for the reader to compare to.

As is, if I'm reading right it looks like these tests would all pass if `applyEvent` just always did nothing :slightly_smiling_face:

Reply:
Hmm I guess the tests in `lib/model/message_test.dart` end up covering those happy paths.

I think it's clearest if the success cases and error cases have their tests next to each other — as I said above, the success tests make a helpful baseline for reading the error tests, and also the reader of the success tests may get nervous if it looks like there are no tests for the error cases.

I like the layer that the `lib/model/message_test.dart` tests are expressed at: calling `store.handleEvent` with a Zulip server event, just as if the event had come in from the server in the polling loop. That's how we try to write most of our model tests, following this principle:
https://zulip.readthedocs.io/en/latest/testing/philosophy.html#integration-testing-or-unit-testing

So let's try to move these error tests to that file and convert them to that file's style.

In fact let's go a step farther: let's move even the "parse poll …" test cases the same way. Then this whole "Message.poll" test group gets replaced by a comment saying the tests for deserializing [Message.poll] can be found over there.

Really what's going on here is that for this data structure, we've fused the job of deserializing it with the job of setting up a nontrivial data structure, including applying some significant model logic to apply all those poll-event submessages. So the deserialization code has a lot of the nature of model code, and it's fitting if its tests therefore wind up in the model code's test files. And then the tests for later events coming in definitely belong on the model side (as discussed above), and they're exercising most of the same code as the deserialization… so the deserialization tests belong next to the later-events tests, in `test/model/`.